alternator/executor.cc

```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
     // try again. Note that simply a missing key is *not* an error (we already
     // handled it above), but this case does include things like timeouts,
     // unavailable CL, etc.
-    return when_all_succeed(response_futures.begin(), response_futures.end()).then(
-            [] (std::vector<std::tuple<std::string, std::optional<rjson::value>>> responses) {
+    return when_all(response_futures.begin(), response_futures.end()).then(
+            [request_items = std::move(request_items)] (std::vector<future<std::tuple<std::string, std::optional<rjson::value>>>> responses) mutable {
         rjson::value response = rjson::empty_object();
         rjson::add(response, "Responses", rjson::empty_object());
         rjson::add(response, "UnprocessedKeys", rjson::empty_object());
-        for (auto& t : responses) {
-            if (!response["Responses"].HasMember(std::get<0>(t).c_str())) {
-                rjson::add_with_string_name(response["Responses"], std::get<0>(t), rjson::empty_array());
+        // In case of full failure (no reads succeeded), an arbitrary error
+        // from one of the operations will be returned.
+        bool some_succeeded = false;
+        std::exception_ptr eptr;
+        // These iterators are used to match keys from the requests with their corresponding responses.
+        // If any of the responses failed, the key iterator will be used to insert
+        // an entry into the UnprocessedKeys object, which will ultimately be returned
+        // to the user in case of partial success of BatchGetItem operation.
+        rjson::value::MemberIterator item_it = request_items.MemberBegin();
+        rjson::value::ValueIterator key_it = item_it->value["Keys"].Begin();
+
+        for (auto& fut : responses) {
+            if (fut.failed()) {
+                eptr = fut.get_exception();
+                if (!response["UnprocessedKeys"].HasMember(item_it->name)) {
+                    rjson::add_with_string_name(response["UnprocessedKeys"], rjson::to_string_view(item_it->name), rjson::empty_object());
+                    rjson::value& unprocessed_item = response["UnprocessedKeys"][item_it->name];
+                    for (auto it = item_it->value.MemberBegin(); it != item_it->value.MemberEnd(); ++it) {
+                        if (it->name != "Keys") {
+                            rjson::add_with_string_name(unprocessed_item, rjson::to_string_view(it->name), rjson::copy(it->value));
```

Comment:
I'm afraid I don't understand this. If we get a failure for any one of the requests from this table, we copy the entire Keys array - possibly many different requested keys - to UnprocessedKeys? Why? Shouldn't it be set here to an empty_array(), and below you copy to it the one failed key? (same as we do below for the sucessful case!)

Reply:
The exact opposite! :) Note that the comparison here is `!=`. The format of `UnprocessedKeys` is super verbose, probably because it's expected to be copied later by the user and reused as input to BatchGetItem. In order to adhere with what DynamoDB returns, I need to copy the entire BatchGetItem metadata **except** keys (projection expression, etc.). Then, I add actual unprocessed keys to the structure one by one - creation of the new array happens in line 3175 and the addition in 3177.

Reply:
Oh, I missed that...